Refactor util/mount interface in prep for moving out of k/k

codenrhoden opened the pull request:
**What this PR does / why we need it**:
This patch refactors `pkg/util/mount` to be more usable outside of
Kubernetes. This is done by refactoring `mount.Interface` to only contain
methods that are not K8s specific. Methods that are not relevant to
basic mount activities but still have OS-specific implementations are
now found in a `mount.HostUtils` interface.

This is in preparation to migrate `pkg/util/mount` out of k/k, where it can be
more easily used by other storage related code, such as CSI drivers.

**Which issue(s) this PR fixes**:
partially addresses #64953

**Special notes for your reviewer**:
Take a close look at `mount.Interface` and `mount.HostUtils` interfaces to see if this is the correct separation.

This was originally 1 big PR that did everything, but was then broken into smaller pieces. All those smaller pieces have been merged.

- #74393
- #74734
- #76047
- #76049
- #76456

**Release Note**:
```release-note
NONE
```

msau42 commented:
/ok-to-test
/assign

You can use hack/update-bazel.sh to automatically make the BUILD file updates

davidz627 commented:
thanks for working on this! I'll take a look soon.
I would suggest that the kubernetes specific mount packages stay as `pkg/util/mount` while the stuff thats going to go out of tree "soon" be called something else, maybe `externalmount` (thats not a great name I wouldn't recommend it) because from the kubernetes perspective the `k8smount` package is just the normal `mount` package.

/assign

codenrhoden commented:
Thanks @davidz627, your naming approach makes sense. Not sure what a better name than `externalmount` is, but I'll keep the K8s specific stuff at `mount`. I should perhaps change the title here to include WIP, as I was waiting to get feedback on the approach before doing all the needed changes.

mcrute commented:
/cc @mcrute

codenrhoden commented:
I pushed some unfinished changes just to make sure the branch was up to date. Still having some specific in-tree volume providers to update, then work on failing tests. But this updates things such that the K8s specific mount count remains in a `mount` package, and the external stuff is now in `externalmount` (sorry, couldn't come up with a better name). Will keep working on this!

codenrhoden commented:
/retest

codenrhoden commented:
/test pull-kubernetes-e2e-gce

codenrhoden commented:
/test pull-kubernetes-local-e2e-containerized
/test pull-kubernetes-e2e-kops-aws

codenrhoden commented:
/skip pull-kubernetes-e2e-kops-aws

codenrhoden commented:
/test pull-kubernetes-local-e2e-containerized

andrewsykim commented:
If the intent is to be able to move `k8s.io/kubernetes/pkg/util/externalmount` to `k8s.io/utils`, `pkg/util/externalmount` shouldn't depend back on `k8s.io` packages (with `k8s.io/klog` and `k8s.io/utils` being the exception for now). There's a dependency to apimachinery there that needs to be removed.

codenrhoden commented:
@andrewsykim That is indeed the intent. Looks like there's a couple of them, actually, but nothing that would be hard to remove. I can definitely satisfy that restriction. Good catch!

dims commented:
/area cloud-provider

@codenrhoden is it possible to pull this off quickly? we could try to get this into 1.14

codenrhoden commented:
@dims yeah I can turn around the changes requested by @andrewsykim and rebase pretty quickly. My main concern has been lack of reviews so far on whether the external library has the right interface. But hoping to get those soon.

dims commented:
/kind cleanup

msau42 commented on pkg/util/externalmount/mount.go:
```diff
@@ -0,0 +1,372 @@
+	// on the system, i.e. still mounted.
+	DeviceOpened(pathname string) (bool, error)
+	// PathIsDevice determines if a path is a device.
+	PathIsDevice(pathname string) (bool, error)
```
This is not used by any plugin and can stay out of this interface

msau42 commented on pkg/util/externalmount/mount.go:
```diff
@@ -0,0 +1,372 @@
+	IsLikelyNotMountPoint(file string) (bool, error)
+	// DeviceOpened determines if the device is in use elsewhere
+	// on the system, i.e. still mounted.
+	DeviceOpened(pathname string) (bool, error)
```
This is used by operation_generator and scaleio plugin, but no other plugins.  I wonder if it's really necessary and if we can modify the scaleio plugin to not use it.

I would like to keep this out of the interface for now until more plugins need to use it.

msau42 commented on pkg/util/externalmount/mount.go:
```diff
@@ -0,0 +1,372 @@
+	GetDeviceNameFromMount(mountPath, pluginDir string) (string, error)
+	// GetFileType checks for file/directory/socket/block/character devices.
+	// Will operate in the host mount namespace if kubelet is running in a container
+	GetFileType(pathname string) (FileType, error)
```
I think we can keep this out for now.  It's only used by hostpath and local plugins which will stay in-tree for now

msau42 commented on pkg/util/externalmount/mount.go:
```diff
@@ -0,0 +1,372 @@
+
+// TODO(thockin): This whole pkg is pretty linux-centric.  As soon as we have
+// an alternate platform, we will need to abstract further.
+package externalmount
```
Can we keep this pkg name as `mount`, and rename the internal one to something like `volume`?  The methods that are left in the internal interface are not really mount related anymore.  I know it will be confusing with `pkg/volume/util`. Maybe we can move everything into the same directory/pkg?  I'm ok with doing it in a separate change to try to keep this smaller.

codenrhoden replied:
yeah, so I did create a new `pkg/util/volume`, which as you said is a bit confusing. But I think merging that with `pkg/volume/util` should be a (much desired) second step.

dims commented:
@codenrhoden can we try to get this into 1.14 please? 

/milestone v1.14

andrewsykim commented:
@msau42 (sorry if I'm missing context) was there any reason for have an internal/external mount package? Is moving the whole mount package out to an external repo still an option?

msau42 commented:
My understanding is that the staging repo is for sharing dependencies  between external and internal components.  In this case, external CSI drivers can benefit from the mount utilities that we've developed.  But things that are only specific to internal kubernetes logic does not need to be exported publicly.

Over time the mount pkg has grown to be more than just mount utilities, and includes a lot of k8s specific logic like subpath and containerized kubelet. Nothing outside of k/k needs this logic, so I think it's cleaner to keep these in tree.  It also makes the mount interface less confusing to external viewers.

andrewsykim commented:
^ sorry I deleted my last comment cause I got my answer from reading the PR history. :P 

For context, my original question was: "What was the reason for internal/external packages? Can we just move the whole package out?" 

Thanks @msau42

andrewsykim commented:
Thoughts on `genericmount` vs `externalmount`?

msau42 commented:
> Thoughts on `genericmount` vs `externalmount` for packge name?

I'm fine with keeping the pkg as "mount".  I'm actually suggesting to rename the remaining internal pkg to something like "volume" because the methods that remain are not actually mount specific.

andrewsykim commented:
Ahh I see, sorry I missed your comment here https://github.com/kubernetes/kubernetes/pull/68513#discussion_r254514789. So if I'm understanding you correctly, it seems like we can break this into separate PRs in this way:
1) extract volume/file specific methods to it's own interface (but we can leave it in the same package to keep the PR smaller). 
2) extract out volume/file interface into it's own package (either `pkg/util/volume` or maybe somewhere in `pkg/volume/util`?)
3) migrate `pkg/util/mount` into it's own external repo (maybe `k8s.io/utils/mount` or `k8s.io/kubernetes-csi/mount`?)

msau42 commented:
That sounds like a good plan.  So maybe this PR should just handle 1)?

andrewsykim commented:
@msau42 works for me, based on your comments the new interface should be:
 
```go
// updated Mount interface
type Interface interface {
	Mount(source string, target string, fstype string, options []string) error
	Unmount(target string) error
	List() ([]MountPoint, error)
	IsMountPointMatch(mp MountPoint, dir string) bool
	IsNotMountPoint(file string) (bool, error)
	IsLikelyNotMountPoint(file string) (bool, error)
	DeviceOpened(pathname string) (bool, error)
	GetDeviceNameFromMount(mountPath, pluginDir string) (string, error)
	MakeRShared(path string) error
	ExistsPath(pathname string) (bool, error)
	CleanSubPaths(podDir string, volumeName string) error
	PrepareSafeSubpath(subPath Subpath) (newHostPath string, cleanupAction func(), err error)
	GetMountRefs(pathname string) ([]string, error)
	GetFSGroup(pathname string) (int64, error)
}
```

With the following moved as exported functions (in the same package), a new interface or removed if there are no callers:
```
MakeFile(pathname string) error
MakeDir(pathname string) error
SafeMakeDir(subdir string, base string, perm os.FileMode) error
PathIsDevice(pathname string) (bool, error)
GetSELinuxSupport(pathname string) (bool, error)
EvalHostSymlinks(pathname string) (string, error)
GetFileType(pathname string) (FileType, error)
GetMode(pathname string) (os.FileMode, error)
```

I think ExistsPath should also be removed

codenrhoden commented:
Thanks for the comments, @msau42 and @andrewsykim ! I'll work on this the next couple of days and get you something new for review.

msau42 commented:
I think these should also be removed from the mount interface:
```
	DeviceOpened(pathname string) (bool, error)
	MakeRShared(path string) error
	CleanSubPaths(podDir string, volumeName string) error
	PrepareSafeSubpath(subPath Subpath) (newHostPath string, cleanupAction func(), err error)
	GetFSGroup(pathname string) (int64, error)
```

The reason why I kept `ExistsPath` in is if you want to consider moving mount_helper.go as well, which will need this.

As for `GetDeviceNameFromMount`, this is a mess and we should try to clean it up in a separate pr first?  There's a mount.`GetDeviceNameFromMount` and a mounter.`GetDeviceNameFromMount`, and different plugins choose different versions.

andrewsykim commented:
> The reason why I kept `ExistsPath` in is if you want to consider moving mount_helper.go as well, which will need this.

Sorry if I'm missing some context, it seems like `mount_helper.go` doesn't use ExistsPath from mounter, but instead uses [PathExists](https://github.com/kubernetes/kubernetes/blob/master/pkg/util/mount/mount_helper.go#L35). Should we still keep ExistsPath in the mount interface? 

Most other uses of `PathExists` seems like it should be replaced with [path.Exists](https://github.com/kubernetes/utils/blob/master/path/file.go#L42-L62). 

> As for `GetDeviceNameFromMount`, this is a mess and we should try to clean it up in a separate pr first? There's a mount.`GetDeviceNameFromMount` and a mounter.`GetDeviceNameFromMount`, and different plugins choose different versions.

^ agreed, I think we should consolidate. Any preference on which implementation we should stick with?

msau42 commented:
> it seems like `mount_helper.go` doesn't use ExistsPath from mounter, but instead uses [PathExists](https://github.com/kubernetes/kubernetes/blob/master/pkg/util/mount/mount_helper.go#L35). Should we still keep ExistsPath in the mount interface?

ugh sorry another instance of confusing naming :(  Yes, in that case, I think ExistsPath can be removed from the mount interface

Re: `GetDeviceNameFromMount`, I took a look at the callers again.  It's used for implementing `ConstructVolumeSpec`, which is not a function that CSI drivers implement.  So let's move it out of the mount interface and file a separate issue to clean it up.

andrewsykim commented:
Thanks for clarifying! So mount interface should just be:

```go
type Interface interface {
	Mount(source string, target string, fstype string, options []string) error
	Unmount(target string) error
	List() ([]MountPoint, error)
	IsMountPointMatch(mp MountPoint, dir string) bool
	IsNotMountPoint(file string) (bool, error)
	IsLikelyNotMountPoint(file string) (bool, error)
	GetMountRefs(pathname string) ([]string, error)
}
```

With the following moved as exported functions (in the same package), a new interface or removed if there are no callers:
```
MakeFile(pathname string) error
MakeDir(pathname string) error
SafeMakeDir(subdir string, base string, perm os.FileMode) error
PathIsDevice(pathname string) (bool, error)
GetSELinuxSupport(pathname string) (bool, error)
EvalHostSymlinks(pathname string) (string, error)
GetFileType(pathname string) (FileType, error)
GetMode(pathname string) (os.FileMode, error)
ExistsPath(pathname string) (bool, error) (replace with https://github.com/kubernetes/utils/blob/master/path/file.go#L42-L62) 
GetDeviceNameFromMount(mountPath, pluginDir string) (string, error)
DeviceOpened(pathname string) (bool, error)
MakeRShared(path string) error
CleanSubPaths(podDir string, volumeName string) error
PrepareSafeSubpath(subPath Subpath) (newHostPath string, cleanupAction func(), err error)
GetFSGroup(pathname string) (int64, error
```

codenrhoden commented:
I started working on this last night, and want to ask a few questions before I go too much farther...

I made the "new" package still be called `mount`, and used the following as `mount.Interface`:

```go
type Interface interface {
	Mount(source string, target string, fstype string, options []string) error
	Unmount(target string) error
	List() ([]MountPoint, error)
	IsMountPointMatch(mp MountPoint, dir string) bool
	IsNotMountPoint(file string) (bool, error)
	IsLikelyNotMountPoint(file string) (bool, error)
	GetMountRefs(pathname string) ([]string, error)
}
```

For all the extra stuff, I created a new interface called `HostUtils`. I stuck with an interface because it needs to be implemented for both `linux` and `windows`, and it's easier to enforce that all implementers do that correctly. The resulting interface looked like this:

```go
type HostUtils interface {
	DeviceOpened(pathname string) (bool, error)
	EvalHostSymlinks(pathname string) (string, error)
	MakeRShared(path string) error
	PathIsDevice(pathname string) (bool, error)
	SafeMakeDir(subdir string, base string, perm os.FileMode) error
	PrepareSafeSubpath(mounter Interface, subPath Subpath) (newHostPath string, cleanupAction func(), err error)
	CleanSubPaths(mounter Interface, podDir string, volumeName string) error
	GetFSGroup(pathname string) (int64, error)
	GetSELinuxSupport(pathname string) (bool, error)
}
```

Of the methods in this interface, `PrepareSafeSubpath` and `CleanSubPaths` are the most complicated. I had to modify the method signature to accept `mount.Interface` because the Linux implementation requires access to a mounter in order to do bind mounds and unmounts when necessary. The approach is a bit K8s specific in my mind, which is why I had originally not moved these two methods to the new package. I might suggest keeping these two methods in the (newly-renamed) `volume` package, or moving them to the `mount` interface so that a `mount.Interface` doesn't have to be passed to them.

I made these methods just be exported functions:

```go
MakeDir(pathname string) error
MakeFile(pathname string) error
GetFileType(pathname string) (FileType, error)
GetDeviceNameFromMount(mounter Interface, mountPath string) (string, int, error)
```

It was pretty easy to just use the common/exported `GetDeviceNameFromMount` and not use the one with multiple implementations, keeping it out of the interface. In fact, the exact same thing can be done with `IsNotMountPoint` if we want, as it exists as both part of the interface, and as an exported function.

I wanted to post all this because once I set all this up, there isn't really anything left in the "old" (newly renamed) `pkg/util/volume`. The only thing that would be left is NSEnter mounter implementation, and an Exec mounter implementation. This is because I moved `mount_helper.go` as well. @msau42 Does this sound like what you want?  I can push a branch for you to take a peek if you want (haven't done it on this PR because there is no point in triggering testing yet).

msau42 commented:
Thanks for the update! I think this sounds good to me.

I am ok moving the subpath methods into its own package: https://github.com/kubernetes/kubernetes/issues/72290.  I don't think it belongs in the mount interface because it's very K8s specific.

We can consider combining/merging the remaining methods into pkg/volume/util, but we could take multiple steps to get there in order to keep the reviews smaller.

dims commented:
/assign @thockin 

Tim, here's a refactoring PR you may be interested in. the idea is to refactor in-place in k/k before moving it out.

codenrhoden commented on pkg/util/mount/mount_helper.go:
```diff
@@ -22,14 +22,16 @@ import (
 // IsNotMountPoint will be called instead of IsLikelyNotMountPoint.
 // IsNotMountPoint is more expensive but properly handles bind mounts within the same fs.
-func CleanupMountPoint(mountPath string, mounter Interface, extensiveMountPointCheck bool) error {
+func CleanupMountPoint(mountPath string, mounter extmount.Interface, extensiveMountPointCheck bool) error {
```
In the newest incarnation, subpaths are  in `pkg/util/volume`. Should be easy to move them to a new package from there.

codenrhoden commented on pkg/util/externalmount/mount.go:
```diff
@@ -0,0 +1,372 @@
+const (
+	// Default mount command if mounter path is not specified
+	defaultMountCommand           = "mount"
+	MountsInGlobalPDPath          = "mounts"
```
I moved it into `pkg/volume/util`.

codenrhoden commented on pkg/util/externalmount/mount.go:
```diff
@@ -0,0 +1,372 @@
+	GetFileType(pathname string) (FileType, error)
+	// MakeFile creates an empty file.
+	// Will operate in the host mount namespace if kubelet is running in a container
+	MakeFile(pathname string) error
```
`MakeFile` and `MakeDir` are now exported functions found in `mount_helper.go`. `SafeMakeDir` is in the new `HostUtils` interface.

codenrhoden commented on pkg/util/externalmount/mount.go:
```diff
@@ -0,0 +1,372 @@
+	ExistsPath(pathname string) (bool, error)
+	// EvalHostSymlinks returns the path name after evaluating symlinks.
+	// Will operate in the host mount namespace if kubelet is running in a container.
+	EvalHostSymlinks(pathname string) (string, error)
```
Moved it to `HostUtils`. Though I have concerns about this that I will explain shortly.

codenrhoden commented on pkg/util/externalmount/mount.go:
```diff
@@ -0,0 +1,372 @@
+	// SELinux.
+	GetSELinuxSupport(pathname string) (bool, error)
+	// GetMode returns permissions of the path.
+	GetMode(pathname string) (os.FileMode, error)
```
I started to do that, but found it is called in `pkg/kubelet/kubelet_pods.go`, so it is now an exported function in `mount_helpers.go` because the implementation was the same for Windows and Linux.

codenrhoden commented:
So while I know tests are going to fail, I wanted to put out a new version that at least compiles and has the right structure...  Testing brought up some issues, though, that I need some help to address.

To recap...

`mount.Interface` now looks like this:

```go
type Interface interface {
	Mount(source string, target string, fstype string, options []string) error
	Unmount(target string) error
	List() ([]MountPoint, error)
	IsMountPointMatch(mp MountPoint, dir string) bool
	IsNotMountPoint(file string) (bool, error)
	IsLikelyNotMountPoint(file string) (bool, error)
	GetMountRefs(pathname string) ([]string, error)
}
```

I created a new `mount.HostUtils` interface that has a lot of the old things from `mount.Interface` that still have OS-specific implementations. That looks like:

```go
type HostUtils interface {
	DeviceOpened(pathname string) (bool, error)
	EvalHostSymlinks(pathname string) (string, error)
	MakeRShared(path string) error
	PathIsDevice(pathname string) (bool, error)
	SafeMakeDir(subdir string, base string, perm os.FileMode) error
	GetFSGroup(pathname string) (int64, error)
	GetSELinuxSupport(pathname string) (bool, error)
}
```

Anything else is now just an exported function (meaning that it could be common against all OS/implementations). 

One nice thing about this approach is the number of modified files dropped from 144 to 65!

My main concern I am seeing right now is anything that uses `NSEnter`, or otherwise expects to access a "mount plugin" on the host rather than what's in a container for containerized kubelet. Without having things like `EvalHostSymlinks`, `GetFSGroup`, `GetMode` etc in `mount.Interface`, I think it will break NSEnter and containerized kubelets, unless we also had a way to register the new "HostUtils" plugins.

As another example, if you look at the changes in `pkg/volume/util/operationexector`, they concern me when it is dealing with containerized kubelet.

@msau42

msau42 commented:
We need to keep the nsenter implementation for `HostUtils` too.  That means we may need two methods to initialize the environment: one for mounter and one for hostutils.

cc @jsafrane for containerized kubelet review

dims commented:
@codenrhoden i am asking the sig-node to deprecate the `containerized kubelet` so we should be ok. Seems like no one is using it and it's not being tested at all outside of the local-e2e-containerized. i have items lined up on the sig-node agenda and PRs as well to drop code related to testing that

msau42 commented:
Even if we deprecate it, I think we still need to support it following the deprecation policy

codenrhoden commented:
> Even if we deprecate it, I think we still need to support it following the deprecation policy

I was going to bring up the same thing. That would be at least two releases out, right?

@msau42 Looking through things a bit closer, I'm unsure how to have the `NSEnter` implementations as part of `HostUtils` because all the existing code expects all the calls where `NSEnter` does things a bit differently to be a part of `mount.Interface`. I think the cleanest way would be to have the kubelet register a `hostutil` plugin just like it does a `mounter` plugin, so that it can use an `NSEnter` version when needed, but that's a more major change than I think people are expecting. Do you see a better way to do it?

codenrhoden commented:
The other way I can see doing it, especially since `containerized kubelet` is being deprecated, is to keep the special implementations of `hostutils` within the `NSenter` mounter, as the were previously, but everywhere that `NSenter` mounter is used within kubelet gets a check to see if it is containerized, and if so, calls that version instead. It's a bit more hackish, as you end with a lot of `if-else` statements, but it keeps the new package cleaner, and once `containerized kubelet` is officially removeds, the `if-else` conditions can be removed as well.

msau42 commented:
I'm not sure I can envision the second approach. Wouldn't hostutils need to be initialized explicitly for linux and windows already?

codenrhoden commented:
> I'm not sure I can envision the second approach. Wouldn't hostutils need to be initialized explicitly for linux and windows already?

@msau42 I was more thinking that in this approach, `hostutils` wouldn't be used directly. All of the `NSenter` methods that had special ways of handling things would stay implemented as `NSenter` methods (essentially, making `NSenter` implement `mount.HostUtils`, just like it used to implement `mount.Interface`), and then calling those when we know `NSenter` had special cases.  So resulting in (pseudo-)code like this:

```
if Containerized:
    mount.MakeRShared()
else:
    hu.MakeRShared()
```

not saying this is a great idea, just trying to think out loud.

as for explicitly initializing HostUtils for Windows/Linux, that would happen on every `mount.NewHostUtils()` anyways, depending on what the binary including the `mount` package is compiled for.

msau42 commented:
I guess what I'm a little confused about, is if we need to call `mount.NewHostUtils()` for linux/windows already, couldn't we also detect and initialize the nsenter implementation in there too?

codenrhoden commented:
You could make an  NSEnter implementation for hostutils, but how do you know when to use it? Maybe I'm missing something.  Right now, the kubelet has a mounter, which implements `mount.Interface`, and then makes calls like `mounter.MakeRShared()`.

For containerized kubelet, it just so happens that `mounter` is actually an `NSEnter` mounter, which does things a bit differently than the standard Linux mounter.

With the new `mount.Interface`, `MakeRShared` is no longer there, so what gets called instead?

I suppose we could make `mount.NewHostUtils()` take a flag that indicates that things are containerized, and then it could do things differently. That would work. But then I still think we need to create a long-lived `hostutils` object that is initialized alongside with the `mounter` that the kubelet can use later on. That might work.

msau42 commented:
Can it be initialized in kubelet the same way nsenter mounter is initialized? https://sourcegraph.com/github.com/kubernetes/kubernetes/-/blob/cmd/kubelet/app/server.go#L373

codenrhoden commented:
@msau42 I believe it can, yes. I've pushed a (non-working) [branch](https://github.com/codenrhoden/kubernetes/compare/mount-refactor-new...codenrhoden:mount-refactor-new-nsenter) so you can see what the differences look like. I'm going to keep working on it.

Moving some of the NSEnter logic into `HostUtils` means adding a few things that were newly common (as exported functions) are going back into the `mount.HostUtils` interface, but I think that's okay.

The main upcoming hurdle I see is that some of the functions within `mount.Interface` will now need access to a `HostUtils` object. The main example of this is there are lots of places where `EvalHostSymlinks` is called, and that implementation is very different when `NSenter` is used. So essentially the `mount.Interface` object and the `mount.HostUtils` object are coupled, kind of defeating the purpose of splitting the original interface into two. I'll keep exploring this and see how it shakes out.

timothysc commented:
@saad-ali and @msau42 - this is currently blocking for other ongoing efforts so IMO this is a priority invert to ~P1.

msau42 commented:
@codenrhoden I took a quick glance at your branch, the initialization is ok to me. My only suggestion is to also have an nsenter implementation of hostutils instead of having "ifs" everywhere in the hostutils_linux.

Regarding the dependency issue, can we have the nsenter initialization also initialize the nsenter hostutils internally, instead of needing to pass it through the mount interface?

codenrhoden commented:
@msau42 yes, I think I see a way to do what you are asking. Hope to have it up for review tomorrow.

codenrhoden commented:
@msau42 Okay, I've push some new commits that illustrate where i think this can go. The NSEnter stuff is now preserved, and I think this is looking pretty good.

I'm hitting a test issue that highlighted the difference between `mount.GetDeviceNameFromMount` and `mounter.GetDeviceNameFromMount`. Unfortunately they do actually behave differently, and the RBD plugin tests highlight this. See #57744 and it's related fix for some details.

Hitting that made me take pause and think about how much is changing here at once, and I'm concerned about breaking a storage plugin in a way that unit tests don't really reveal. So I'm trying to think about how to break this PR into smaller pieces.

One first step to make this one smaller is to refactor subpath into it's own package, ala #72290. In fact, this PR already does all that work, so I might as well split that out and get that merged first.

As for the two versions of `GetDeviceNameFromMount`, I'm inclined to leave it on this pass so we don't break anything. Smaller steps. :) Thoughts?

codenrhoden commented:
I opened PR #74393 to cover pulling subpath out into it's own package. That part will be much easier to review on its own.

msau42 commented:
Thanks, that sounds like a good plan to me.  Can you open a separate issue for the GetDeviceNameFromMount issue, and we can track that separately?

codenrhoden commented:
> Thanks, that sounds like a good plan to me. Can you open a separate issue for the GetDeviceNameFromMount issue, and we can track that separately?

Done. #74548

codenrhoden commented:
For anyone following this PR, PTAL #74393. It is a subset of this PR, and is ready to go!

codenrhoden commented:
@msau42 Now that #74393 has merged, I'm trying to think of additional small chunks I can take on to make this larger  PR easier. I was originally thinking of moving the NSEnter mounter into `k8s.io/util/nsenter`, but that can't happen until `mount.interface` is also in `k8s.io/util`. As part of this PR, I had moved that mounter from `pkg/util/mount` to `pkg/util/volume`, but I'd rather find a better place for it. That would be an easy thing to do separately.

Where do you think is a good place for the NSEnter mounter to live? `pkg/volume/util/nsenter` ? `pkg/volume/util` ?

msau42 commented:
`pkg/volume/util/nsenter` sounds ok to me.

How about for now, focus on moving out the HostUtil methods?  Also it would be nice to come up with a better name, but I don't really have any great ideas. PathUtil? FileUtil?

codenrhoden commented:
> How about for now, focus on moving out the HostUtil methods? Also it would be nice to come up with a better name, but I don't really have any great ideas. PathUtil? FileUtil?

@msau42 Okay, I can do that. I will make that this PR, as it's really a refactoring of `mount.Interface`. I'll try to pare down from what I did before (e.g. not moving `nsenter` yet) to make the review for this easier and the PR smaller. Can always pick those things off in smaller pieces.

I also don't know what to call the collective Interface that contains what we take out of `mount.Interface`. PathUtil would be confused with `k8s.io/utils/path`. We are really looking at an interface like:

```go
type HostUtils interface {
	// DeviceOpened determines if the device is in use elsewhere
	// on the system, i.e. still mounted.
	DeviceOpened(pathname string) (bool, error)

	// EvalHostSymlinks returns the path name after evaluating symlinks.
	EvalHostSymlinks(pathname string) (string, error)

	// Will operate in the host mount namespace if kubelet is running in a container
	// MakeRShared checks that given path is on a mount with 'rshared' mount
	// propagation. If not, it bind-mounts the path as rshared.
	MakeRShared(path string) error

	// PathIsDevice determines if a path is a device.
	PathIsDevice(pathname string) (bool, error)

	// GetFSGroup returns FSGroup of the path.
	GetFSGroup(pathname string) (int64, error)

	// GetSELinuxSupport returns true if given path is on a mount that supports
	// SELinux.
	GetSELinuxSupport(pathname string) (bool, error)
}
```

These are all utilities that have to do with files/paths/devices. While I originally had tried to pull it out to an exported function, `EvalHostSymlinks` stays in the interface because it is implemented much differently for `NSEnter`.

I'll make the fewest number of changes I can here. And try to think of something better than `HostUtils`. =) One question -- do you picture `HostUtils` staying in the mount package (even when it moves to `k8s.io/utils`, becoming a new package, or staying in k/k? I think some of it's basic functionality could move into `k8s.io/utils/path` now that I think about it. But that's all stuff for later. Or maybe not. Maybe that's the better way to do it -- pick off a couple things here that can be moved straight to `k8s.io/utils/path`?

Definitely food for thought. But regardless, I have enough to take a new stab at this.

soggiest commented:
1.14 code freeze is coming in 3 days. It looks like this won't be making it into 1.14. Can we remove the milestone?

andrewsykim commented:
/milestone v1.15

codenrhoden commented:
@msau42 I've opened #76047 as another part of breaking this into smaller pieces. The new PR deals with moving the NSEnter mounter to a new path.

I know we talked about focusing on refactoring out the `HostUtils` piece next, but I'm starting to question why to go through the effort of refactoring those pieces out of `mount.Interface`. Taking the subpath pieces out made since, because that stays in-tree. But pulling out the `HostUtils` part (or whatever we call it) is creating two interfaces that both have to be implemented, and are generally both used at the same time in the same places. I'm looking to see if I can find a clean example where this is *not* the case in order to justify the additional work.

codenrhoden committed:
Refactor pkg/util/mount to be more reusable

This patch refactors pkg/util/mount to be more usable outside of
Kubernetes. This is done by refactoring mount.Interface to only contain
methods that are not K8s specific. Methods that are not relevant to
basic mount activities but still have OS-specific implementations are
now found in a mount.HostUtils interface.

codenrhoden commented:
/test pull-kubernetes-integration

codenrhoden commented:
/test pull-kubernetes-e2e-kops-aws
/test pull-kubernetes-local-e2e-containerized

codenrhoden commented:
Hi @msau42,

I now have a clear version of this PR again! Original PR was a +4900/-4700 code change, now this is more manageable at +650/-600. Part of that is because I still didn't go as far as before -- while the `HostUtils` stuff is still there, I didn't move anything out into its own file just yet. That's stuff that could be done *after* the code is moved if we want, and now it is easier to review and has no worries about losing history.

Current mount interfaces are defined as so:

```go
type Interface interface {
	// Mount mounts source to target as fstype with given options.
	Mount(source string, target string, fstype string, options []string) error
	// Unmount unmounts given target.
	Unmount(target string) error
	// List returns a list of all mounted filesystems.  This can be large.
	// On some platforms, reading mounts is not guaranteed consistent (i.e.
	// it could change between chunked reads). This is guaranteed to be
	// consistent.
	List() ([]MountPoint, error)
	// IsMountPointMatch determines if the mountpoint matches the dir
	IsMountPointMatch(mp MountPoint, dir string) bool
	// IsLikelyNotMountPoint uses heuristics to determine if a directory
	// is a mountpoint.
	// It should return ErrNotExist when the directory does not exist.
	// IsLikelyNotMountPoint does NOT properly detect all mountpoint types
	// most notably linux bind mounts.
	IsLikelyNotMountPoint(file string) (bool, error)
	// GetMountRefs finds all mount references to the path, returns a
	// list of paths. Path could be a mountpoint path, device or a normal
	// directory (for bind mount).
	GetMountRefs(pathname string) ([]string, error)
}

type HostUtils interface {
	// DeviceOpened determines if the device is in use elsewhere
	// on the system, i.e. still mounted.
	DeviceOpened(pathname string) (bool, error)
	// PathIsDevice determines if a path is a device.
	PathIsDevice(pathname string) (bool, error)
	// GetDeviceNameFromMount finds the device name by checking the mount path
	// to get the global mount path within its plugin directory
	GetDeviceNameFromMount(mounter Interface, mountPath, pluginMountDir string) (string, error)
	// MakeRShared checks that given path is on a mount with 'rshared' mount
	// propagation. If not, it bind-mounts the path as rshared.
	MakeRShared(path string) error
	// GetFileType checks for file/directory/socket/block/character devices.
	// Will operate in the host mount namespace if kubelet is running in a container
	GetFileType(pathname string) (FileType, error)
	// MakeFile creates an empty file.
	// Will operate in the host mount namespace if kubelet is running in a container
	MakeFile(pathname string) error
	// MakeDir creates a new directory.
	// Will operate in the host mount namespace if kubelet is running in a container
	MakeDir(pathname string) error
	// Will operate in the host mount namespace if kubelet is running in a container.
	// Error is returned on any other error than "file not found".
	ExistsPath(pathname string) (bool, error)
	// EvalHostSymlinks returns the path name after evaluating symlinks.
	// Will operate in the host mount namespace if kubelet is running in a container.
	EvalHostSymlinks(pathname string) (string, error)
	// GetFSGroup returns FSGroup of the path.
	GetFSGroup(pathname string) (int64, error)
	// GetSELinuxSupport returns true if given path is on a mount that supports
	// SELinux.
	GetSELinuxSupport(pathname string) (bool, error)
	// GetMode returns permissions of the path.
	GetMode(pathname string) (os.FileMode, error)
}
```

I'd like to try to iterate on this quickly so we can get it in and the move it to its final destination with plenty of time before code freeze.

codenrhoden commented on pkg/kubelet/kubelet_test.go:
```diff
@@ -311,7 +312,6 @@ func newTestKubeletWithImageList(
 		NewInitializedVolumePluginMgr(kubelet, kubelet.secretManager, kubelet.configMapManager, token.NewManager(kubelet.kubeClient), allPlugins, prober)
 	require.NoError(t, err, "Failed to initialize VolumePluginMgr")
 
-	kubelet.mounter = &mount.FakeMounter{}
```
I removed this because it is a redefinition -- it was already set at L163

msau42 commented on pkg/volume/plugins.go:
```diff
@@ -412,6 +412,9 @@ type VolumeHost interface {
 	GetHostIP() (net.IP, error)
 
+	// Returns HostUtils Interface
+	GetHostUtil() mount.HostUtils
```
We're trying to start separating out VolumeHost into controller-specific interfaces.  Since this is only used by kubelet, can you add this to KubeletVolumeHost?

codenrhoden commented:
/test pull-kubernetes-local-e2e-containerized
/refresh

codenrhoden commented:
/test pull-kubernetes-local-e2e-containerized

codenrhoden commented:
The `pull-kubernetes-local-e2e-containerized` job was removed from test-infra back in February. =)

soggiest commented:
Hello! Code Freeze is just about on us. We'll be entering into Freeze tomorrow, May 31st. Is this still planned to land in 1.15?

msau42 commented on pkg/volume/awsebs/aws_ebs.go:
```diff
@@ -250,8 +250,9 @@ func getVolumeSource(
 
 func (plugin *awsElasticBlockStorePlugin) ConstructVolumeSpec(volName, mountPath string) (*volume.Spec, error) {
 	mounter := plugin.host.GetMounter(plugin.GetPluginName())
+	hu := plugin.host.(volume.KubeletVolumeHost).GetHostUtil()
```
what happens if the casting fails?

codenrhoden replied:
Good point, I will do this safely instead. =)

msau42 commented on pkg/volume/plugins.go:
```diff
@@ -334,6 +334,8 @@ type KubeletVolumeHost interface {
 	// WaitForCacheSync is a helper function that waits for cache sync for CSIDriverLister
 	WaitForCacheSync() error
+	// Returns HostUtils Interface
+	GetHostUtil() mount.HostUtils
```
Is this used?

codenrhoden replied:
Yes. In the other place you commented on -- it's used throughout various volume plugins.

codenrhoden commented:
@msau42 updated based on your comments. 👍

msau42 commented on pkg/volume/csi/csi_mounter.go:
```diff
@@ -328,9 +328,18 @@ func (c *csiMountMgr) podAttributes() (map[string]string, error) {
-	path := c.GetPath()
-	supportSelinux, err := mounter.GetSELinuxSupport(path)
+	kvh, ok := c.plugin.host.(volume.KubeletVolumeHost)
+	var (
```
The ordering of this is strange. Can you put the vars on top?

msau42 commented on pkg/volume/csi/csi_mounter.go:
```diff
@@ -328,9 +328,18 @@ func (c *csiMountMgr) podAttributes() (map[string]string, error) {
+		err            error
+	)
+	if !ok {
+		klog.V(2).Info(log("plugin volume host does not implement KubeletVolumeHost interface"))
```
This seems like a fatal error if it fails.  Can we try to initialize host util earlier and pass it in through csiMountMgr?

codenrhoden replied:
@msau42 I updated as you requested. the `csiMountMgr` struct is just created directly inside of `csi_mounter.go`, instead of a constructor of some type, so it's not quite as clean as I would have liked.  There are a few other places in `csi_mounter.go` where the same cast is done, and they don't treat it as fatal. If they had all treated it as fatal, I would have just done the cast once and saved it inside of the `csiPlugin` struct instead.

msau42 replied:
I think checking in NewMounter() and NewUnmounter() are fine.

msau42 commented:
/lgtm
/approve

Thanks so much for taking on this huge refactor!

msau42 commented:
/assign @jsafrane

codenrhoden commented:
Awesome! I'm sorry I didn't get it in for 1.15... With KubeCon right before code freeze, I think reviewers were busy with many other things. :) I'll wrangle all the other needed approvers once code freeze lifts. Excited to move forward on this one.

codenrhoden commented:
/test pull-kubernetes-bazel-test

codenrhoden commented:
/test pull-kubernetes-e2e-gce

jsafrane commented:
/lgtm
/approve

thockin commented on pkg/util/mount/fake.go:
```diff
@@ -29,7 +29,6 @@ import (
 type FakeMounter struct {
```
Fakes should go in a testing/ sub-pkg

codenrhoden replied:
I don't disagree. Since that part of the code wasn't modified as part of this PR, I'd be happy to try and restructure that in a follow-on PR, most likely after this package is moved out of k/k.

codenrhoden commented:
/assign @wojtek-t

derekwaynecarr commented:
/approve
/lgtm

wojtek-t commented:
/approve

codenrhoden commented:
/test pull-kubernetes-bazel-build
